README
Plack-Middleware-CombineStatic version 0.03
===========================================

The README is used to introduce the module and provide instructions on
how to install the module, any machine dependencies it may have (for
example C compilers and installed libraries) and any other information
that should be provided before the module is installed.

A README file is required for CPAN modules since CPAN extracts the
README file from a module distribution so that people browsing the
archive can use it get an idea of the modules uses. It is usually a
good idea to provide version information here so that people can
decide whether fixes for the module are worth downloading.

INSTALLATION

To install this module type the following:

   perl Makefile.PL
   make
   make test
   make install

DEPENDENCIES

This module requires these other modules and libraries:

  blah blah blah

COPYRIGHT AND LICENCE

Put the correct copyright and licence information here.

Copyright (C) 2012 by tjmc

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself, either Perl version 5.15.9 or,
at your option, any later version of Perl 5 you may have available.




Changes
Revision history for Perl extension Plack::Middleware::DynamicAssets.

0.01  Tue May 15 21:35:56 2012
	- original version
		-X Plack::Middleware::DynamicAssets



lib/Plack/Middleware/CombineStatic.pm
package Plack::Middleware::CombineStatic;

use parent 'Plack::Middleware';

use strict;
use warnings;
use Plack::Util::Accessor qw(minifiers parameter root cache path);
use Plack::Request;
use Plack::MIME;
use Path::Class qw(dir);
use Digest::MD5 'md5_base64';
use Try::Tiny;

our $VERSION = '0.3';


=head1 NAME Plack::Middleware::CombineStatic


=head1 SYNOPSIS

    use Plack::Builder;
    use Plack;:Middleware::CombineStatic;

    my $app = sub { ... };

    my $cache = Some::Cache->new();

    builder {
        enable 'Plack::Middleware::CombineStatic',
            root => '/var/www/',
            parameter => 'filez',
            path   => qr{\.js|\.css},
            cache  => $cache,
            minifiers => {
                'application/javascript'=> sub { JavaScript::Minifier::XS::minify(@_) },
            };

        $app;
    };


    # in the browser
    GET http://localhost:5000/?filez=one.js,two.js,three.js
    
    # or 
    GET http://localhost:5000/?filez=three.js,two.js,five.js

    # etc.
=cut

sub call {
    my $self = shift;
    my $env  = shift;

    my $request = Plack::Request->new($env);

    my $ret = try {

        my ( $content_type, $content );
        if ( my $cached = $self->_get_from_cache($request) ) {
            ( $content_type, $content ) = @$cached;
        }
        else {

            my @files = split /,/,
              ( $request->param( $self->parameter ) || '' );

            my $path_info = $request->path_info;

            return unless @files;
            return if grep { $_ !~ $self->path } @files;

            my @paths = map { dir( $self->root, $path_info )->file($_) } @files;

            ( $content_type, $content ) = $self->_slurp(@paths);

            $content = $self->_minify( $content_type, $content );

            $self->_set_in_cache( $request, [ $content_type, $content ] );

        }

        if ($content) {

            return [
                200,
                [
                    'Content-Length' => length($content),
                    'Content-Type'   => $content_type,
                ],
                [$content],
            ];
        }
    }
    catch {
        return $self->_fault($_);
    };

    return $ret if $ret;

    return $self->app->($env);
}

sub _fault {

    my $self = shift;

    my $fault_message = shift || '[FAILED TO LOAD STATIC RESOURCE]';

    return [
        500,
        [
            'Content-Length' => length($fault_message),
            'Content-Type'   => 'text/plain',
        ],
        [$fault_message],
    ];
}

sub _slurp {
    my $self = shift;

    my ( $content_type, $content );

    for my $file (@_) {
        $content_type ||= Plack::MIME->mime_type($file);

        die "mixing different file types isn't allowed"
          if $content_type ne Plack::MIME->mime_type($file);

        $content .= $file->slurp;
    }

    return $content_type, $content;

}

sub _minify {
    my $self = shift;
    my ( $content_type, $content ) = @_;

    return $content unless exists $self->minifiers->{$content_type};

    my $minifier = $self->minifiers->{$content_type};

    return $minifier->($content);
}

sub _get_from_cache {
    my ( $self, $request ) = @_;

    return unless $self->cache;

    my $cache_key = md5_base64( $request->param( $self->parameter ) );
    return $self->cache->get($cache_key);
}

sub _set_in_cache {
    my ( $self, $request, $data ) = @_;

    return unless $self->cache;

    my $cache_key = md5_base64( $request->param( $self->parameter ) );
    return $self->cache->set( $cache_key, $data );
}

=head1 SEE ALSO

L<Plack::Middleware::Static|Plack::Middleware::Static>

L<Plack::Middleware::Static::Minifier|Plack::Middleware::Static::Minifier>

L<Plack::Middleware::Assets|Plack::Middleware::Assets>



=head1 AUTHOR

tjmc

=cut

1;


lib/Plack/Middleware/DynamicAssets.pm
package Plack::Middleware::DynamicAssets;

use parent 'Plack::Middleware';

use strict;
use warnings;
use Plack::Util::Accessor qw(minify parameter root cache);
use Plack::Request;
use Plack::MIME;
use Perl6::Slurp;
use Class::Load;
use Digest::MD5 'md5_base64';
use Try::Tiny;

our $VERSION = '0.2';

our %minifiers = (
    'javascript' => 'JavaScript::Minifier::XS',
    'css'        => 'CSS::Minifier::XS',
);

=head1 NAME Plack::Middleware::DynamicAssets


=head1 SYNOPSIS

    use Plack::Builder;
    use Plack;:Middleware::DynamicAssets;

    my $app = sub { ... };

    my $cache = Some::Cache->new();

    builder {
        enable 'Plack::Middleware::DynamicAssets',
            root => '/var/www/',
            parameter => 'filez',
            cache  => $cache,
            minify => 1;

        $app;
    };


    # in the browser
    GET http://localhost:5000/?filez=one.js,two.js,three.js
    
    # or 
    GET http://localhost:5000/?filez=three.js,two.js,five.js

    # etc.
=cut

sub call {
    my $self = shift;
    my $env  = shift;

    my $request = Plack::Request->new($env);

    my $ret = try {

        my ( $content_type, $content );
        if ( my $cached = $self->_get_from_cache($request) ) {
            ( $content_type, $content ) = @$cached;
        }
        else {

            my @files = split /,/, $request->param( $self->parameter );
            my @paths = map { join '/', $self->root, $_ } @files;

            ( $content_type, $content ) = $self->_slurp(@paths);

            $content = $self->_minify( $content_type, $content );
            $self->_set_in_cache( $request, [ $content_type, $content ] );

        }

        if ($content) {

            return [
                200,
                [
                    'Content-Length' => length($content),
                    'Content-Type'   => $content_type,
                ],
                [$content],
            ];
        }
    }
    catch {
        return $self->_fault;
    };

    return $ret if $ret;

    return $self->app->($env);
}

sub _fault {

    my $fault_message = '[FAILED TO LOAD STATIC RESOURCE]';
    return [
        500,
        [
            'Content-Length' => length($fault_message),
            'Content-Type'   => 'text/plain',
        ],
        [$fault_message],
    ];
}

sub _slurp {
    my $self = shift;

    my ( $content_type, $content );

    for my $file (@_) {
        $content_type ||= Plack::MIME->mime_type($file);

        die "mixing different file types isn't allowed"
          if $content_type ne Plack::MIME->mime_type($file);

        $content .= slurp $file;
    }

    return $content_type, $content;

}

sub _minify {
    my $self = shift;
    my ( $content_type, $content ) = @_;

    return $content unless $self->minify();

    my $type = $content_type =~ /javascript/ ? 'javascript' : 'css';

    Class::Load::load_class( $minifiers{$type} );

    if ( $type eq 'javascript' ) {
        return JavaScript::Minifier::XS::minify($content);
    }
    else {
        return CSS::Minifier::XS::minify($content);
    }
}

sub _get_from_cache {
    my ( $self, $request ) = @_;

    return unless $self->cache;

    my $cache_key = md5_base64( $request->param( $self->parameter ) );
    return $self->cache->get($cache_key);
}

sub _set_in_cache {
    my ( $self, $request, $data ) = @_;

    return unless $self->cache;

    my $cache_key = md5_base64( $request->param( $self->parameter ) );
    return $self->cache->set( $cache_key, $data );
}

=head1 SEE ALSO

L<Plack::Middleware::Static|Plack::Middleware::Static>

L<Plack::Middleware::Static::Minifier|Plack::Middleware::Static::Minifier>

L<Plack::Middleware::Assets|Plack::Middleware::Assets>



=head1 AUTHOR

tjmc

=cut

1;
